### PR TITLE
[AND-3223] Improve handling of cause of load/play of Ad fails

### PR DESCRIPTION
**Pull request by HuFangshuai**
PR for ticket: https://vungle.atlassian.net/browse/AND-3223

1. removed canPlayAd check.
2. Added error code mapping.

### CODE REVIEW

File: Vungle/src/main/java/com/mopub/mobileads/VungleInterstitial.java [hunk 1]
```diff
@@ -288,69 +282,88 @@ public void run() {
...
+            case AD_FAILED_TO_DOWNLOAD:
+            case ASSET_DOWNLOAD_ERROR:
+                return VIDEO_DOWNLOAD_ERROR;
+            case AD_UNABLE_TO_PLAY:
+                return VIDEO_PLAYBACK_ERROR;
+            default:
+                return UNSPECIFIED;
+        }
+    }
+
```
Comment (anton-blashchuk) [inline Vungle/src/main/java/com/mopub/mobileads/VungleInterstitial.java]: it duplicates code in VungleBanner. maybe we should either make one of them package-visible and use from both classes and optionally move to some third class in package(leaving package visibility)

**Review by anton-blashchuk — APPROVED** [review]
one minor issue, can be merged as is or slightly updated

File: Vungle/src/main/java/com/mopub/mobileads/VungleRouter.java [hunk 2]
```diff
@@ -47,10 +47,10 @@ public void onResume(@NonNull final Activity activity) {
             super.onResume(activity);
         }
     };
-    private static VungleRouter sInstance = new VungleRouter();
+    private static final VungleRouter sInstance = new VungleRouter();
     private static SDKInitState sInitState = SDKInitState.NOTINITIALIZED;
-    private static Map<String, VungleRouterListener> sVungleRouterListeners = new HashMap<>();
-    private static Map<String, VungleRouterListener> sWaitingList = new HashMap<>();
+    private static final Map<String, VungleRouterListener> sVungleRouterListeners = new HashMap<>();
+    private static final Map<String, VungleRouterListener> sWaitingList = new HashMap<>();
```
Comment (anton-blashchuk) [inline Vungle/src/main/java/com/mopub/mobileads/VungleRouter.java]: one more thing - if we're making these `final`, they should be renamed, - INSTANCE, WAITING_LIST, VUNGLE_ROUTER_LISTENERS

File: Vungle/src/main/java/com/mopub/mobileads/VungleBanner.java [hunk 3]
```diff
@@ -347,136 +350,157 @@ public void run() {
...
-                            mLoadListener.onAdLoadFailed(MoPubErrorCode.NETWORK_NO_FILL);
-                            MoPubLog.log(getAdNetworkId(), LOAD_FAILED, ADAPTER_NAME,
-                                    MoPubErrorCode.NETWORK_NO_FILL.getIntCode(),
-                                    MoPubErrorCode.NETWORK_NO_FILL);
-                        } else if (mInteractionListener != null) {
-                            mInteractionListener.onAdFailed(MoPubErrorCode.NETWORK_NO_FILL);
-                            MoPubLog.log(getAdNetworkId(), SHOW_FAILED, ADAPTER_NAME,
-                                    MoPubErrorCode.NETWORK_NO_FILL.getIntCode(),
-                                    MoPubErrorCode.NETWORK_NO_FILL);
+                        if (mInteractionListener != null) {
```
Comment (HuFangshuai) [inline Vungle/src/main/java/com/mopub/mobileads/VungleBanner.java:347]: Don't know why adding mLoadListener related code. For Interstitial ad, find the same issue in VungleInterstitial.java file as well.
Related PR: https://github.com/mopub/mopub-android-mediation/pull/257/commits/86d1c3e0f9469c32520d097f9456048ddb8fac1a#diff-aceb4d76c06d9382480e8f101358698c3e8c041110a4e1c7d93efd5d88d967a7R352-R362

File: Vungle/src/main/java/com/mopub/mobileads/VungleInterstitial.java [hunk 4]
```diff
@@ -288,69 +282,88 @@ public void run() {
...
-                            mLoadListener.onAdLoadFailed(MoPubErrorCode.NETWORK_NO_FILL);
-                            MoPubLog.log(getAdNetworkId(), LOAD_FAILED, ADAPTER_NAME,
-                                    MoPubErrorCode.NETWORK_NO_FILL.getIntCode(),
-                                    MoPubErrorCode.NETWORK_NO_FILL);
-                        } else if (mInteractionListener != null) {
-                            mInteractionListener.onAdFailed(MoPubErrorCode.NETWORK_NO_FILL);
-                            MoPubLog.log(getAdNetworkId(), SHOW_FAILED, ADAPTER_NAME,
-                                    MoPubErrorCode.NETWORK_NO_FILL.getIntCode(),
-                                    MoPubErrorCode.NETWORK_NO_FILL);
+                        if (mInteractionListener != null) {
```
Comment (HuFangshuai) [inline Vungle/src/main/java/com/mopub/mobileads/VungleInterstitial.java:280]: The same issue about mLoadListener logic: https://github.com/mopub/mopub-android-mediation/commit/83d908ba389cc6c2f931c416b2341cdb00d57bc7#diff-c14a90e6291945340ba12d6896ee4aebe86e46a9ee70ecc31a757d287da8e41b

File: Vungle/src/main/java/com/mopub/mobileads/VungleRewardedVideo.java [hunk 5]
```diff
@@ -322,47 +337,73 @@ public void onAdViewed(@NonNull String placementReferenceId) {
...
             if (mPlacementId.equals(placementReferenceId)) {
-                MoPubLog.log(getAdNetworkId(), CUSTOM, ADAPTER_NAME, "onUnableToPlayAd - Placement ID: " +
-                        placementReferenceId + ", reason: " + reason);
+                MoPubLog.log(getAdNetworkId(), CUSTOM, ADAPTER_NAME, "onAdPlayError - Placement ID: " +
+                        placementReferenceId + ", reason: " + error.getLocalizedMessage());
 
                 mIsPlaying = false;
+                MoPubErrorCode errorCode = mapErrorCode(error.getExceptionCode());
+
                 if (mInteractionListener != null) {
```
Comment (HuFangshuai) [inline Vungle/src/main/java/com/mopub/mobileads/VungleRewardedVideo.java:336]: We can find there's no mLoadListener related code here for Rewarded ad.

**Review by anton-blashchuk — APPROVED** [review]
still lgtm :)

**Pull request merged**